### ISSUE TITLE
Request to generate initial code on project creation

### ISSUE THREAD

**Issue by nagesh007** [processing/processing4#898]
### Relevant sub-area for this feature?

PDE

### Feature description

Whenever we click on "File > New", New sketch appears but **it will be empty**. We need to start writing setup and draw functions manually. I kindly request you to generate initial code on new sketch creation. (Maybe this feature can switched on/off in preferences)

### Benefits

- Developer can easily start writing his logic without writing basic functions **every time**

**Comment by SableRaf** [comment]
Thank you for the suggestion! One of the nice things about Processing is that even a single line of code is a valid program. There’s no need for setup or draw unless you want to change the sketch size or animate something. For this reason, we don’t plan to make this change.

**Comment by Stefterv** [comment]
Agreed but maybe we could allow users to define their starting sketch for themselves, like blender does with File->Defaults->Save Startup File

**Comment by hx2A** [comment]
I think there's a good idea here. What if the "Sketch" menu had an "insert template" submenu option, with a few choices like "basic animated sketch", that would insert something like this:

```java
void setup() {
  size(500, 500);
  // write code that will be called once in this function
}

void draw() {
  // write code that will be called for every frame here
}
```

and maybe an "interactive sketch" option that had the above template plus more empty functions for the mouse and key events. And the comments in the template could match the user's language preferences.

This would be helpful for beginners because it would provide some guidance on code structure and also save them the time of looking up the names of the mouse and key event functions when they forget. And it would help me because I keep starting my Processing Sketches with "def setup()" by mistake...

**Comment by SableRaf** [comment]
Good points both of you. I read a bit too quickly and didn't notice @nagesh007's suggestion to make it an optional setting. My apologies!

**Comment by SushantBansal-tech** [comment]
@SableRaf Can you please assign this issue to me

**Comment by SableRaf** [comment]
Hi @SushantBansal-tech, I just assigned you. Thanks for picking this up! Let us know if you need any guidance. How are you thinking of approaching this issue?

**Comment by SushantBansal-tech** [comment]
@SableRaf I have just check out the sketch.java and sketchCode.java in which newfile variable  has made and is using PApplet in this
`File newFile = new File(folder, newName);
`
` if (renamingCode) {
      if (currentIndex == 0 &&
          Preferences.getBoolean("editor.sync_folder_and_filename")) {
        if (!renameSketch(newName, newExtension)) return;

      } else {  // else if something besides code[0], or ok to decouple name
        if (!current.renameTo(newFile, newExtension)) {
          Messages.showWarning(Language.text("name.messages.error"),
                               Language.interpolate("name.messages.no_rename_file.description",
                               current.getFileName(), newFile.getName()));
          return;
        }
        if (currentIndex == 0) {
          // If the main tab was renamed, check sketch.properties
          mainFile = newFile;  //code[0].getFile();
          updateNameProperties();
        }
      }

    } else {  // not renaming, creating a new file
      try {
        if (!newFile.createNewFile()) {
          // Already checking for IOException, so make our own.
          throw new IOException("createNewFile() returned false");
        }
      } catch (IOException e) {
        Messages.showWarning(Language.text("name.messages.error"),
                             Language.interpolate("name.messages.no_create_file.description",
                             newFile, folder.getAbsolutePath()), e);
        return;
      }
      SketchCode newCode = new SketchCode(newFile, newExtension);
      insertCode(newCode);
    }

    // sort the entries
    sortCode();

    // set the new guy as current
    setCurrentCode(newName);

    // update the tabs
    editor.rebuildHeader();
  }`
 `  protected void insertCode(SketchCode newCode) {
    // make sure the user didn't hide the sketch folder
    ensureExistence();
    System.out.println("InsertCode here newCode");

    // add file to the code/codeCount list, resort the list
    //if (codeCount == code.length) {
    code = (SketchCode[]) PApplet.append(code, newCode);
    codeCount++;
    //}
    //code[codeCount++] = newCode;
  }`  In this i will add a default code having settings() and setup() and draw function() .
Please let me Know if I am going in the right way

**Comment by SushantBansal-tech** [comment]
@SableRaf . Please provide some guidance  I have discussed some thing about how I am approaching this issue

**Comment by SushantBansal-tech** [comment]
@SableRaf  I want to modify the Processing source code to automatically insert some default code whenever a new tab is created. However, I am having trouble understanding the interaction between Sketch.java, SketchCode.java, Editor.java, and PApplet.java.

Could you guide me on:

The execution order when a new tab is created.

Where I should modify the code to insert default content into a new tab.

How these classes interact when handling multiple tabs.

**Comment by SableRaf** [comment]
@SushantBansal-tech I'm leaning more towards the suggestion by @hx2A to have a new menu item for inserting a template. 

> What if the "Sketch" menu had an "insert template" submenu option, with a few choices like "basic animated sketch"

cc @Stefterv

**Comment by SushantBansal-tech** [comment]
@SableRaf okay I understand what we have to do 
I will create a new menu in which certain templates will be  provided

**Comment by Stefterv** [comment]
My personal preference would still be to define a file for a new sketch (shocker!), in any case please also think about how users of the PDE will be able to define their own templates to be inserted.

For the first iteration I think it is fine if those templates just get _pasted_ into the editor tab you are looking at, later we can look at making that interaction a bit more refined if it needs it UX wise

**Comment by SableRaf** [comment]
That makes sense to me!

We can start with a predefined set of maybe three templates in the menu. It's easiest to implement and gives us something useful right away. 

Then we can look into the following improvements:

1. A preference option in Settings to pick a default template to use every time you hit “New...”
3. A `templates` folder in the sketchbook to store user-created templates.
5. A menu item to save the current sketch as a custom template.

How does that sound?

**Comment by SushantBansal-tech** [comment]
@SableRaf Should I create a new menu in the menubar or insert as an menuitem in a sketch menu???

![Image](https://github.com/user-attachments/assets/78eb50e2-358a-4138-9190-24755a4c205f)

**Comment by SableRaf** [comment]
Thanks for your question @SushantBansal-tech!

I'd suggest adding a `File > New From Template` menu item with a selection of a few basic templates. 

Let's start with the following templates:

**Animation Sketch**
```processing
void setup() {
  size(500, 500);
  // write code that will be called once in this function
}

void draw() {
  // write code that will be called for every frame here
}
```

**Interactive Sketch**
```processing
void setup() {
  size(500, 500);
  // write code that will be called once in this function
}

void draw() {
  // write code that will be called for every frame here
}

void mousePressed() {
  ellipse(mouseX, mouseY, 20, 20);
  // write code that will run when you click
}
```

**Fullscreen Sketch**
```processing
void setup() {
  fullScreen();  // create a fullscreen canvas
}

void draw() {
  circle(width / 2, height / 2, height * 0.5);
}
```

**Resizeable Sketch**
```processing
void setup() {
  size(500, 500);
  windowResizable(true);
  // allow the window to be resized
}

void draw() {
  circle(width / 2, height / 2, min(width, height) * 0.5);
  // draw a circle that resizes with the window
}

void windowResized() {
  println("Window resized to: " + width + "x" + height);
  // this function is called whenever the window is resized
}
```

**Comment by SushantBansal-tech** [comment]
@SableRaf what about the Examples in file Menu ?? It is also providing the code for different stuff.  do we actually want this only ?? or something different from this ..

![Image](https://github.com/user-attachments/assets/370222da-19d4-4d1b-94b5-4bcfc2623b97)

**Comment by SableRaf** [comment]
Hi @SushantBansal-tech 

The focus of templates as proposed in this issue is to make it easier to start a new sketch with a basic structure. 

Examples serve a different need of demonstrating Processing’s features. So let's keep those two things separate for now.

I suggest we continue with the “New From Template” submenu.

**Comment by SushantBansal-tech** [comment]
@SableRaf  I have implemented the above features.

![Image](https://github.com/user-attachments/assets/5c48bb3c-f4a8-4446-8274-62fdff706226)

![Image](https://github.com/user-attachments/assets/caef2484-746e-489e-a2aa-67b73682374b)
for this I have make alter the code in editor.java and make a new file Template.java

**Comment by SushantBansal-tech** [comment]
@SableRaf Should I raise the PR

**Comment by SushantBansal-tech** [comment]
@SableRaf  Can You please review My PR [1020](https://github.com/processing/processing4/pull/1020)

**Comment by nikhilreddy006** [comment]
Hi everyone! I have an idea for this feature. In addition to adding default setup() and draw() functions when creating a new sketch, I propose including console.log("Hello World") by default. This would provide a clear starting point for new users and help them see immediate feedback in the console.

What do you think? I’d love to hear feedback and explore if this could be a useful addition for beginners!

![Image](https://github.com/user-attachments/assets/cc48ccc2-b484-4487-a137-4f52b3cfe97a)

**Comment by Vaivaswat2244** [comment]
Hello everyone @SableRaf @Stefterv , was going through this issue. Really interesting idea. I would love to make this work. 
Can this be assigned?
Also I saw the pr made be @SushantBansal-tech . The issue which is mentioned there regarding `insertText` function. Was that fixed or should we go through that first?

**Comment by SableRaf** [comment]
Hi @Vaivaswat2244, thanks for your interest in this issue!

After testing the earlier PR, we found that implementing this properly would likely mean reworking how modes are structured. That’s a bigger change than we can take on right now.

We’re putting it on hold for now, but we still think the idea is worth exploring and hope to come back to it later. If you have ideas in the meantime, feel free to share them here.
